pkg/events/derive/hidden_kernel_module.go

```diff
@@ -25,69 +26,188 @@ import (
+	wakeupChannel            = make(chan ScanRequest)
 )
 
+const (
+	procModules    uint32 = 1 << 0
+	kset                  = 1 << 1
+	modTree               = 1 << 2
+	NewMod                = 1 << 3
+	SendToVerifier        = 1 << 29
+	FullScan              = 1 << 30
+	HiddenModule          = 1 << 31
+)
+
+type ScanRequest struct {
+	Address uint64
+	Flags   uint32
+}
+
 func HiddenKernelModule() DeriveFunction {
 	return deriveSingleEvent(events.HiddenKernelModule, deriveHiddenKernelModulesArgs())
 }
 
+// deriveHiddenKernelModulesArgs
 func deriveHiddenKernelModulesArgs() deriveArgsFunction {
 	return func(event trace.Event) ([]interface{}, error) {
 		address, err := parse.ArgVal[uint64](&event, "address")
 		if err != nil {
 			return nil, err
 		}
 
-		name, err := parse.ArgVal[[]byte](&event, "name")
+		if _, found := foundHiddenKernModsCache.Get(address); found {
+			return nil, nil // Already reported this event - no need to report again
+		}
+
+		flags, err := parse.ArgVal[uint32](&event, "flags")
 		if err != nil {
 			return nil, err
 		}
 
-		if _, found := foundHiddenKernModsCache.Get(address); found {
-			return nil, nil // already reported this event - no need to report again
-		} else {
-			foundHiddenKernModsCache.Add(address, struct{}{}) // so we won't report multiple times
+		var nameBytes []byte
+		var name string
+
+		if flags&HiddenModule != 0 {
+			// Continue to submit the event to the user!
+		} else if flags&procModules != 0 {
+			// Check again against /proc/modules to address possible race condition
+			nameBytes, err = parse.ArgVal[[]byte](&event, "name")
+			if err != nil {
+				return nil, err
+			}
+			name = string(nameBytes[:bytes.IndexByte(nameBytes[:], 0)]) // Remove the trailing terminating characters
+
+			isFound, err := recheckProcModules(name)
+			if err != nil {
+				return nil, err
+			}
+
+			if isFound {
+				return nil, nil // It's not hidden
+			}
+		} else if flags&SendToVerifier != 0 {
+			// Do another check for this particular module against the modules list once again to tackle race conditions.
+			// If it still isn't found, we'll receive this event once again with flags that will cause it to be sent to user
+			wakeupChannel <- ScanRequest{Address: address, Flags: flags}
+			return nil, nil
+		} else if flags&FullScan != 0 {
+			wakeupChannel <- ScanRequest{Flags: flags} // No need to send the address as we're doing a full generic scan
+			return nil, nil
+		} else if flags&NewMod != 0 {
+			startScanTime := address // The address field in this case is unused, so it's being used as the start scan time
+			err := newModsCheckForHidden(startScanTime, flags)
+			return nil, err
+		}
+
+		if nameBytes == nil { // If we didn't parse it in procModules condition above, parse it now
+			nameBytes, err := parse.ArgVal[[]byte](&event, "name")
+			if err != nil {
+				return nil, err
+			}
+			name = string(nameBytes[:bytes.IndexByte(nameBytes[:], 0)]) // Remove the trailing terminating characters
+		}
```

Comment:
This is the part I'm referring to as "the logic better documented", so when we have to support a bug here, for example, we don't have to re-read it all over (and can gain couple of hours).

Example: "we will receive this event once again with flags that will case it to be sent to user". What flags ? When will be sent to user ? How ? When will it come back ? Etc...